Clients: simplify adding and removing RSE distances

0xquark opened the pull request:
<!-- Please read https://rucio.cern.ch/documentation/contributing before submitting a pull request -->
Fixes #7570

0xquark committed:
clients: Simplify adding and removing RSE distances; fixes #7570

Updates to `add-distance`:

- Added a `--bidirectional` flag to set distances in both directions between two RSEs.
Example:
rucio-admin rse add-distance --bidirectional --distance 1 RSE1 RSE2

Updates to `delete-distance`:

- Added `--bidirectional` flag to remove distances in both directions at once.

- Now supports using site names (not just RSE names) when deleting distances.

- Added `--src` and `--dest` flags to remove all outgoing or incoming distances from/to a specific site or RSE.

Other changes:

- Help text updated to explain new options.
- Test cases updated to cover the new features.
- Always pass -y flag in non-interactive mode in cli.
- Use mutually exclusive group for `--src` and `--dest`, not allowing them to be passed together.
- Add a check for positional argument in the begining and assign it to `source` and `destination` in case of `src_flag` and `dest_flag` being passed.

Signed-off-by: Karanjot Singh <karanjot.singh@cern.ch>

0xquark committed:
clients: Use ExportClient to fetch RSEs and Distances; Fixes #7570

- ExportClient has `ExportData` method which allows us to get complete export of RSE data with distances. We can then create an internal mesh network to search between the data in order to find the distances and RSEs for source and destination. With this approach, we don't rely on calling the client multiple times to check if it's an RSE or a Site or to find the distances.

Other Changes:
- Improve logging messages when no distance is found for deletion.
- Add type annotations in `distance_remove`

Signed-off-by: Karanjot Singh <karanjot.singh@cern.ch>

0xquark committed:
clients: Add rich config and improve logging messages

- Use `rich` in order to show information/error messages to user.

0xquark commented on tests/test_bin_rucio.py:
```diff
@@ -283,6 +283,11 @@ def test_rse_add_distance(self):
         print(out, err)
         assert exitcode == 0
+        temprse3 = rse_name_generator()
+        cmd = 'rucio-admin rse add %s' % temprse3
```
I noticed `rucio-admin` is still used in the tests, even though it's deprecated. I followed the same pattern for my tests—should I switch to using the new CLI instead? in tests.

voetberg replied:
`rucio-admin` isn't yet deprecated, it's just *going* to be. For now I don't see an issue keeping the old format until it's fully retired

rdimaio replied:
I'd use the new CLI in tests, though. When it will be deprecated, we'll have to migrate the tests to the new syntax, so we might as well simplify that process by writing new tests with the new syntax and without `rucio-admin` - https://github.com/rucio/rucio/issues/7260

0xquark replied:
I believe using the new CLI in tests is a good idea, but my concern is whether we've finalized the commands mentioned in [https://github.com/rucio/rucio/issues/7260](https://github.com/rucio/rucio/issues/7260). I don't want the tests to fail if we decide to change the names of some commands later.

0xquark commented on tests/test_bin_rucio.py:
```diff
@@ -303,6 +308,36 @@ def test_rse_add_distance(self):
+        assert "Set distance from %s to %s to 1" % (temprse3, temprse1) in out
+        
+        # Verify distances were set in both directions
+        cmd = 'rucio-admin rse get-distance %s %s' % (temprse1, temprse3)
```
Is there a reason we only use assertions to check the commands, and not do any cross-verification? I feel like cross-verifying could help catch issues if something's wrong with the gateway or server.

voetberg replied:
We cross check in some tests, but this is generally done through making a second call with the py-client. I believe calls back to the core are rare because a) This ideally is covered by tests for the individual components and b) The client test suite doesn't assume there is a DB connection exposed. 

@rdimaio can correct me if I'm wrong

rdimaio replied:
This is a client-level test, and in my opinion here we should only care that the client works correctly, i.e. getting the expected output for a given input. All the layers called by this client call are expected to be tested individually, to ensure they're all individually working correctly, + the integ tests

voetberg commented on tests/test_bin_rucio.py:
```diff
@@ -303,6 +308,36 @@ def test_rse_add_distance(self):
+        cmd = 'rucio-admin rse add-distance %s %s' % (temprse1, generate_uuid())
+        exitcode, out, err = execute(cmd)
+        print(out, err)
+        assert exitcode != 0
```
`execute` generally only produces an exitcode of 0, 1, 2 - you want to check for 1 here (1 is "failed to successfully complete" and 2 is "failed to parse the CLI input")

0xquark replied:
Shouldn't the test needs to be failed for both the reasons? I mean if we check only for 1 here, we are limiting the reason of failure and making the test more brittle, if the implementation or anything changes in future which results in validation error, it could result error code 2. 
With `assert exitcode != 0` we don't worry about the reason but more about the failure and expanding the scope. let me know what do you think?

voetberg replied:
In principal I don't think you're wrong, but in practice, this can hide problems with the test stemming form the CLI parsing, which is a different sort of check. You should check that it is specifically failing for the reason you expect it to fail - like you can check that the error message calls out that the RSE doesn't exist in addition to the exitcode.

voetberg replied:
Which you do and I missed that it's right below this comment, lol. Nvm.

voetberg commented on tests/test_bin_rucio.py:
```diff
@@ -610,24 +869,21 @@ def test_upload_expiration_date(self):
         exitcode, out, err = execute(cmd)
-        print(out)
-        print(err)
+        print(out, err)
```
This probably shouldn't be included in this PR (it has nothing to do with RSE distance setting)

voetberg commented on lib/rucio/cli/bin_legacy/rucio_admin.py:
```diff
@@ -607,13 +613,207 @@ def update_distance_rses(args, client, logger, console, spinner):
+                    logger.info(f"Deleted distance information from {src} to {dst}")
+                else:
+                    print(f"Deleted distance information from {src} to {dst}")
+            except Exception as e:
```
Catching all exceptions like this overwrites some of the error handling done in `exception_handler`, you should only catch `NoDistance` or `RSENotFound`.

voetberg commented on lib/rucio/cli/rse.py:
```diff
@@ -126,10 +126,12 @@ def distance_add(ctx, source_rse, destination_rse, distance):
 @distance.command("remove")
 @click.argument("source-rse")
 @click.argument("destination-rse")
+@click.option('-y', '--yes', is_flag=True, help='Automatically confirm deletion')
```
You can use a [`confirmation_option`](https://click.palletsprojects.com/en/stable/api/#click.confirmation_option) to streamline this, but I don't know if this `confirmation` functionality is something we want period. I'm a little dubious about this just because I don't know how this works into people's existing workflows

0xquark replied:
I suggest we stick with the current approach of using `click.option` instead of switching to `confirmation_option` because it gives us more control. Right now, we first show which distances will be deleted, then ask for confirmation, and finally delete them.

With confirmation_option, it seems like the confirmation prompt would appear right when the command starts before showing what will be deleted—which isn’t ideal. I haven’t tested it, but that’s how I expect it to behave.

Current flow:
Show distances to delete → Ask for confirmation → Delete

Expected confirmation_option flow:
Ask for confirmation right away → Show distances → Delete (assuming)

Also, I’m not sure if this new option would support different modes like Rich or plain text.

You can see the current deletion confirmation implementation here:
https://github.com/rucio/rucio/blob/fc6fb2d40fcad06d1f0159fc479304791063efeb/lib/rucio/cli/bin_legacy/rucio_admin.py#L766-L776

The above `@click.option('-y', '--yes', is_flag=True, help='Automatically confirm deletion')` is for skipping the confirmation box in the command.

voetberg replied:
You can connect 'confirmation option' to the prompt if you are concerned with showing proper information first (examples here)[https://click.palletsprojects.com/en/stable/prompts/#confirmation-prompts], but I think I disagree with using confirmation prompts o n principal (unless they are included everywhere). It creates an inconsistent user experience. I would like this option removed for now and we can open it up for discussion at a later date.

0xquark replied:
I believe deleting distances in bulk is a major and potentially risky action, so we should always confirm with the user before doing it. Showing a confirmation prompt is a good way to make sure they know exactly what will be deleted. In this case, it’s more important to focus on user safety and clarity than on having a perfectly consistent user experience.

voetberg replied:
I really do think we *should* do that (I tried to add it to the CLI rewrite lol), but we don't. This is a change beyond the stated goal of the issue/pr so for tracking/release management purposes, it should be a different issue.

0xquark replied:
What I suggest is, Let's have the confirmation block in this feature since this is something which was initially asked in the issue (https://github.com/rucio/rucio/issues/7570#issuecomment-2801240422) and open a new issue for major overwrite elsewhere in the cli.

voetberg commented on lib/rucio/cli/rse.py:
```diff
@@ -126,10 +126,12 @@ def distance_add(ctx, source_rse, destination_rse, distance):
+def distance_remove(ctx: click.Context, source_rse: str, destination_rse: str, yes: bool) -> None:
     """Un-link SOURCE-RSE from DESTINATION-RSE by removing the distance between them"""
-    args = Arguments({"no_pager": ctx.obj.no_pager, "source": source_rse, "destination": destination_rse})
+    # Setting yes=True for non-interactive test environments, but keeping bidirectional=False to match legacy behavior
```
I don't understand this comment, why are we limiting to `bidirection=False` for the newer client?

0xquark replied:
Before, the CLI didn’t have a bidirectional option. If we now make a breaking change where bidirectional is set to true, it would delete distances in both directions which we probably don’t want. Earlier, `distance_remove` only deleted the distance in one direction, so the old behavior should stay the same even with the new feature.

voetberg replied:
I understand that, but why not just add an option for bidirectional and default it to False?

0xquark replied:
Now I Understood! Thanks. 
I was confused before with New CLI vs Legacy CLI. So the options for new cli are added using click in `rse.py` while the options for old cli are added through argparser in `bin_legacy/rucio_admin.py`. 
I'll add the `bidirectional` option configurable in the new cli.

voetberg commented on lib/rucio/cli/bin_legacy/rucio_admin.py:
```diff
@@ -607,13 +613,207 @@ def update_distance_rses(args, client, logger, console, spinner):
+        if cli_config == 'rich':
+            distance_tree = Tree(f"The following {len(distances_to_delete)} distances will be deleted:")
+            for src, dst in sorted(distances_to_delete):
+                distance_tree.add(f"{src} → {dst}")
```
I think this is the only place `→` is ever used in rucio, for consistency I'd use `->` instead.

voetberg commented on lib/rucio/cli/bin_legacy/rucio_admin.py:
```diff
@@ -607,13 +613,207 @@ def update_distance_rses(args, client, logger, console, spinner):
+            # Check if source is an RSE
+            if args.source in all_rses:
+                src_rses = [args.source]
+            else:
```
This is the only place in the client were we allow the definition of "rse" to also mean "site at an RSE", I think this is confusing and should not be allowed.

0xquark replied:
I think this is the only place in the client where you can use a site name instead of an RSE name. Everywhere else in the code, when it asks for an RSE, it expects a real RSE name. But here, the input can be either a site or an RSE so you can delete the distance between two specific RSEs, or between all RSEs at one site and all RSEs at another site. I’ll update the comments to explain this more clearly.

voetberg commented on lib/rucio/cli/bin_legacy/rucio_admin.py:
```diff
@@ -1922,18 +2129,38 @@ def get_parser():
-    delete_distance_rses_parser.add_argument(dest='destination', action='store', help='Destination RSE name')
+    
+    # Create a mutually exclusive group for source/destination specification
+    src_dest_group = delete_distance_rses_parser.add_mutually_exclusive_group()
```
This causes more problems than its worth, I think using just `source, nargs=?` `dest, nargs=?`

0xquark replied:
Do you mean to remove the `--src` and `--dest` flags? and instead use source and destination only. 
If we do this, in order to delete all the incoming links to a destination, we would need to specify it using an empty space for the source `rucio-admin rse delete-distance '' JDOE_SCRATCHDISK`. According to me, in terms of user experience, `rucio-admin rse delete-distance --dest JDOE_SCRATCHDISK` is much more better.

voetberg replied:
I completely agree with that, but this additional option of having the both options in mutually exclusively groups is way more confusing than having just one or the other and introduces all that nested logic in the function itself. I would recommend leaving "source"/"destination" in rucio-admin, and moving the '--src' '--dest' flags to the new client in `cli/rse.py`

0xquark replied:
Great! I've moved the flags (`--src` and `--dest`) to be part of the new cli and only kept positional source and destination arguments in the legacy cli.

voetberg commented on lib/rucio/cli/bin_legacy/rucio_admin.py:
```diff
@@ -607,13 +613,207 @@ def update_distance_rses(args, client, logger, console, spinner):
+            spinner.stop()
+        
+        # Handle case when no distances are found
+        if not distances_to_delete:
```
It makes more sense to just raise an error here, the result is the same and the error message is cleaner.

voetberg commented on lib/rucio/cli/bin_legacy/rucio_admin.py:
```diff
@@ -607,13 +613,207 @@ def update_distance_rses(args, client, logger, console, spinner):
+                    for dst in all_distances[src]:
+                        distances_to_delete.append((src, dst))
+        
+        elif args.destination:
```
This just duplicates the source rse selection, no? It would be cleaner to just do `if args.source or args.destination` and handle the logic in the rse loop.

voetberg replied:
Hi, please don't resolve comments until you pushed the change, it makes it harder to follow up

rcarpa commented:
Would be nice to use this new feature to set distances in the integration tests here : https://github.com/rucio/rucio/blob/22868d219b45929a7e3016d12f0d5b67e18166d2/tools/docker_activate_rses.sh#L95

0xquark committed:
suggestions: revert print(out,err)

0xquark committed:
suggestions: Fix arrow usage

0xquark committed:
suggestions: Add better comments for differentiating source and destination RSE

Signed-off-by: Karanjot Singh <karanjot.singh@cern.ch>

0xquark committed:
suggestions: Add bidirectional option in both add_distance and remove_distance for new cli in rse.py

0xquark committed:
suggestions: Refactor delete_distance_rses method to streamline handling of source and destination rses

0xquark committed:
suggestions: Catch NoDistance and RSENotFound exceptions specifically

0xquark committed:
suggestions: Raise NoDistance error when no distances are found

0xquark committed:
ruff: fix linting error

Signed-off-by: Karanjot Singh <karanjot.singh@cern.ch>

0xquark committed:
suggestions: Use --src and --dest flags in the new cli for removing incoming and outgoing distance to an rse; Remove the flags from the old cli and use Source and Destination.

0xquark committed:
fix: Update tests with the new flags

0xquark committed:
fix: fix tests using shell escaping for bash

0xquark committed:
suggestions: Replace the legacy cli with new cli in tests for this feature

0xquark commented:
```
______________________ TestBinRucio.test_rse_add_distance ______________________
[gw1] linux -- Python 3.9.21 /opt/venv/bin/python
tests/test_bin_rucio.py:309: in test_rse_add_distance
    assert exitcode != 0
E   assert 0 != 0
------------------------------ Captured log setup ------------------------------
DEBUG    urllib3.connectionpool:connectionpool.py:1021 Starting new HTTPS connection (1): localhost:443
DEBUG    urllib3.connectionpool:connectionpool.py:474 https://localhost:443 "GET /auth/userpass HTTP/1.1" 200 0
DEBUG    urllib3.connectionpool:connectionpool.py:1021 Starting new HTTPS connection (2): localhost:443
DEBUG    urllib3.connectionpool:connectionpool.py:474 https://localhost:443 "GET /rses/MOCK4 HTTP/1.1" 200 927
DEBUG    urllib3.connectionpool:connectionpool.py:1021 Starting new HTTPS connection (1): localhost:443
DEBUG    urllib3.connectionpool:connectionpool.py:474 https://localhost:443 "POST /accountlimits/local/root/MOCK4 HTTP/1.1" 201 7
DEBUG    urllib3.connectionpool:connectionpool.py:474 https://localhost:443 "POST /rses/MOCK4/attr/istape HTTP/1.1" 201 7
----------------------------- Captured stdout call -----------------------------
Added new deterministic RSE: MOCK-YSQEASDEKE
 
Added new deterministic RSE: MOCK-CGGUTJOJWZ
 
Added new deterministic RSE: MOCK-AEJCMYGSLD
 
$> rucio rse distance add MOCK-YSQEASDEKE MOCK-CGGUTJOJWZ --distance 1
Set distance from MOCK-YSQEASDEKE to MOCK-CGGUTJOJWZ to 1
 
$> rucio rse distance add MOCK-CGGUTJOJWZ MOCK-YSQEASDEKE --distance 1
Set distance from MOCK-CGGUTJOJWZ to MOCK-YSQEASDEKE to 1
 
$> rucio rse distance add MOCK-CGGUTJOJWZ MOCK-YSQEASDEKE --distance 1
 2025-05-12 12:18:29,933	ERROR	An object with the same identifier already exists.
Details: Distance from MOCK-CGGUTJOJWZ to MOCK-YSQEASDEKE already exists!
 0
 ```
 Why does the new cli provides with exit code 0 for an error? I think this is a bug.
 @voetberg

voetberg commented:
> ```
> ______________________ TestBinRucio.test_rse_add_distance ______________________
> [gw1] linux -- Python 3.9.21 /opt/venv/bin/python
> tests/test_bin_rucio.py:309: in test_rse_add_distance
>     assert exitcode != 0
> E   assert 0 != 0
> ------------------------------ Captured log setup ------------------------------
> DEBUG    urllib3.connectionpool:connectionpool.py:1021 Starting new HTTPS connection (1): localhost:443
> DEBUG    urllib3.connectionpool:connectionpool.py:474 https://localhost:443 "GET /auth/userpass HTTP/1.1" 200 0
> DEBUG    urllib3.connectionpool:connectionpool.py:1021 Starting new HTTPS connection (2): localhost:443
> DEBUG    urllib3.connectionpool:connectionpool.py:474 https://localhost:443 "GET /rses/MOCK4 HTTP/1.1" 200 927
> DEBUG    urllib3.connectionpool:connectionpool.py:1021 Starting new HTTPS connection (1): localhost:443
> DEBUG    urllib3.connectionpool:connectionpool.py:474 https://localhost:443 "POST /accountlimits/local/root/MOCK4 HTTP/1.1" 201 7
> DEBUG    urllib3.connectionpool:connectionpool.py:474 https://localhost:443 "POST /rses/MOCK4/attr/istape HTTP/1.1" 201 7
> ----------------------------- Captured stdout call -----------------------------
> Added new deterministic RSE: MOCK-YSQEASDEKE
>  
> Added new deterministic RSE: MOCK-CGGUTJOJWZ
>  
> Added new deterministic RSE: MOCK-AEJCMYGSLD
>  
> $> rucio rse distance add MOCK-YSQEASDEKE MOCK-CGGUTJOJWZ --distance 1
> Set distance from MOCK-YSQEASDEKE to MOCK-CGGUTJOJWZ to 1
>  
> $> rucio rse distance add MOCK-CGGUTJOJWZ MOCK-YSQEASDEKE --distance 1
> Set distance from MOCK-CGGUTJOJWZ to MOCK-YSQEASDEKE to 1
>  
> $> rucio rse distance add MOCK-CGGUTJOJWZ MOCK-YSQEASDEKE --distance 1
>  2025-05-12 12:18:29,933	ERROR	An object with the same identifier already exists.
> Details: Distance from MOCK-CGGUTJOJWZ to MOCK-YSQEASDEKE already exists!
>  0
> ```
> 
> Why does the new cli provides with exit code 0 for an error? I think this is a bug. @voetberg

Ah, I see what is happening here. The legacy CLI doesn't throw an error it just returns an integer 1 to indicate an error, which the error handling doesn't catch because the new CLI doesn't return anything. I suppose that counts as a bug, but I'm not sure how best to go about solving it. 

You can either add a return statement for the new CLI (`return add_distance_rses(...)`, or throw an error (raise an `Duplicate` error) - we don't have a solid design for this so I'll let you decide.

dchristidis reviewed (commented):
I’m afraid both the design and the implementation are heading in the wrong direction.

Design-wise, the way to go from a single RSE to many RSEs is through an RSE expression. This is our construct and it should be used wherever possible. An RSE name is a valid RSE expression, so it’s backwards-compatible. Remember that the `site` RSE attribute is not guaranteed to be set. And the way to express all RSEs is through the RSE expression `*`.

Implementation-wise, all this logic should **EDIT: not** exist in the client. The core/gateway/API should be adapted to accept RSE expressions instead of RSE names.

0xquark committed:
suggestions: Update RSE distance remove to support RSE expressions

- Changed argument names from 'source-rse' and 'destination-rse' to 'source' and 'destination' for clarity.

- Added docs to include examples and support for RSE expressions.

- Updated the 'delete_distance_rses' function to handle RSE expressions for both source and destination.

- Adjusted related tests to reflect the new structure and usage of RSE expressions.

0xquark commented:
It’s been quite a while, and rebasing everything would be a bit challenging, so I’ll open a new PR with the updates. Closing this one now.
